pkg/kv/txn_external_test.go

```diff
@@ -700,6 +717,177 @@ func TestGenerateForcedRetryableErrorByPoisoning(t *testing.T) {
+	}
+	keyA, keyB, keyC := mkKey("a"), mkKey("b"), mkKey("c")
+	// Create a read committed transaction that writes to keys "a" in its first
+	// statement before hitting a retryable error during the second statement.
+	// Roll back the operations performed on behalf of just the second statement
+	// and retry it.
+	txn := kv.NewTxn(ctx, db, 0 /* gatewayNodeID */)
+	// Enable stepping so that we can control statement boudaries.
+	txn.ConfigureStepping(ctx, kv.SteppingEnabled)
+	require.NoError(t, txn.SetIsoLevel(isolation.ReadCommitted))
+	// Write to "a" in the first "statement".
+	stmt1, err := txn.CreateSavepoint(ctx)
+	require.NoError(t, err)
+	require.NoError(t, txn.Put(ctx, keyA, 1))
+	require.NoError(t, txn.ReleaseSavepoint(ctx, stmt1))
+	require.NoError(t, txn.Step(ctx))
+	// Perform a series of reads and writes in the second "statement".
+	stmt2, err := txn.CreateSavepoint(ctx)
+	require.NoError(t, err)
+	// Read from "b" and "c" to establish refresh spans.
+	res, err := txn.Get(ctx, keyB)
+	require.NoError(t, err)
+	require.False(t, res.Exists())
+	res, err = txn.Get(ctx, keyC)
+	require.NoError(t, err)
+	require.False(t, res.Exists())
+	// Write to "c" outside the transaction to create a write-write conflict.
+	require.NoError(t, db.Put(ctx, keyC, 2))
+	// Write to "b" in the transaction, without issue.
+	require.NoError(t, txn.Put(ctx, keyB, 3))
+	// Write to "c" in the transaction to hit the write-write conflict, which
+	// causes the statement to need to retry.
+	err = txn.Put(ctx, keyC, 4)
+	require.Error(t, err)
+	var retryErr *kvpb.TransactionRetryWithProtoRefreshError
+	require.ErrorAs(t, err, &retryErr)
+	require.Contains(t, retryErr.Error(), "WriteTooOldError")
+	require.False(t, retryErr.TxnMustRestartFromBeginning())
+	// Attempts to use the transaction again before calling
+	// TestPrepareForPartialRetry are rejected with the same retry error.
+	_, err = txn.Get(ctx, keyB)
+	require.Error(t, err)
+	require.ErrorAs(t, err, &retryErr)
+	require.Contains(t, retryErr.Error(), "WriteTooOldError")
+	// Roll back the second statement's writes and prepare to retry it.
+	require.NoError(t, txn.RollbackToSavepoint(ctx, stmt2))
+	// PrepareForPartialRetry resets the transaction and allows it to be used
+	// again.
+	require.NoError(t, txn.PrepareForPartialRetry(ctx))
+	// PrepareForPartialRetry is idempotent.
+	require.NoError(t, txn.PrepareForPartialRetry(ctx))
+	// The transaction's own write on "a" from statement 1 was not discarded.
+	res, err = txn.Get(ctx, keyA)
+	require.NoError(t, err)
+	require.Equal(t, int64(1), res.ValueInt())
+	// The transaction's own write on "b" from statement 2 was discarded.
+	res, err = txn.Get(ctx, keyB)
+	require.NoError(t, err)
+	require.False(t, res.Exists())
+	// The non-transactional write on "c" is now visible.
+	res, err = txn.Get(ctx, keyC)
+	require.NoError(t, err)
+	require.Equal(t, int64(2), res.ValueInt())
+	// Write to "b" in the transaction again.
```

Comment:
TODO: change this to write to "d" instead so that we demonstrate that the transaction's write to "b" was not only not visible, but also fully discarded.